bpf/lib/fib.h

```diff
@@ -127,43 +123,37 @@ fib_do_redirect(struct __ctx_buff *ctx, const bool needs_l2_check,
 			return DROP_WRITE_ERROR;
 		break;
 	case BPF_FIB_LKUP_RET_NO_NEIGH:
-		/* previous fib lookup was performed, we can fillout both
-		 * a bpf_redir_neigh and a dmac.
-		 *
-		 * the former is used if we have access to redirect_neigh
-		 * the latter is used if we don't and have to use the eBPF
-		 * neighbor map.
-		 */
-		if (fib_params) {
-			nh_params.nh_family = fib_params->l.family;
-			__bpf_memcpy_builtin(&nh_params.ipv6_nh,
-					     &fib_params->l.ipv6_dst,
-					     sizeof(nh_params.ipv6_nh));
-			nh = &nh_params;
-
-			if (!neigh_resolver_available() && allow_neigh_map) {
-				/* The neigh_record_ip{4,6} locations are mainly from
-				 * inbound client traffic on the load-balancer where we
-				 * know that replies need to go back to them.
-				 */
-				dmac = fib_params->l.family == AF_INET ?
-				neigh_lookup_ip4(&fib_params->l.ipv4_dst) :
-				neigh_lookup_ip6((void *)&fib_params->l.ipv6_dst);
-			}
-		}
-
 		/* If we are able to resolve neighbors on demand, always
 		 * prefer that over the BPF neighbor map since the latter
 		 * might be less accurate in some asymmetric corner cases.
 		 */
 		if (neigh_resolver_available()) {
-			if (nh)
+			if (fib_params) {
+				struct bpf_redir_neigh nh_params;
+
+				nh_params.nh_family = fib_params->l.family;
+				__bpf_memcpy_builtin(&nh_params.ipv6_nh,
+						     &fib_params->l.ipv6_dst,
+						     sizeof(nh_params.ipv6_nh));
+
 				return redirect_neigh(*oif, &nh_params,
 						sizeof(nh_params), 0);
-			else
-				return redirect_neigh(*oif, NULL, 0, 0);
+			}
+
+			return redirect_neigh(*oif, NULL, 0, 0);
 		} else {
 			union macaddr smac = NATIVE_DEV_MAC_BY_IFINDEX(*oif);
+			union macaddr *dmac = NULL;
+
+			if (allow_neigh_map) {
+				/* The neigh_record_ip{4,6} locations are mainly from
+				 * inbound client traffic on the load-balancer where we
+				 * know that replies need to go back to them.
+				 */
+				dmac = fib_params->l.family == AF_INET ?
```

Comment:
Do we need to check `if (fib_params)` here?

Reply:
I think that's something we'll need to investigate in a follow-on PR.

For users of `fib_redirect_v*()`, it gets all handled transparently - if there's no neigh-resolver available, they will always prepare + pass a `fib_params` struct.

"Expert" users that call `fib_do_redirect()` need to have some understanding what's needed for the fall back to the neigh-map. It's not easy to make that fool-proof. They could easily pass a 0-initialized fib_params - it would satisfy the suggested check and would satisfy the verifier, but still not help the neigh-map lookup.